quarkc/javascript.py

```diff
@@ -90,8 +99,12 @@ def name(n):
 def type(path, name, parameters):
     return ".".join(path + [name])
 
-def import_(path, origin, dep):
+def import_(path, origin, dep, seen=None, lazy=False):
     qual = qualify(path, origin)
+    if seen is not None:
```

Comment:
Mutating an incoming data structure feels like an anti-pattern to me, especially since this is only used in one language. Side-effects are bad enough, side-effects involving caller tracking state seems worse, it's breaking abstraction layer.

My thought: future implementation of backends should involve an instance of a class, rather than calling functions on a module. That will allow tracking state within the backend. Even better they can be immutable objects, and then there's no side-effects at all, state changes are expressed by swapping out a new backend for the existing one returned by `import_` and friends.